core/src/main/java/com/linecorp/armeria/common/QueryStringDecoder.java

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright 2012 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.common;
+
+import static io.netty.util.internal.StringUtil.SPACE;
+
+import java.util.Arrays;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.linecorp.armeria.internal.TemporaryThreadLocals;
+
+final class QueryStringDecoder {
+
+    @SuppressWarnings("checkstyle:AvoidEscapedUnicodeCharacters")
+    private static final char UNKNOWN_CHAR = '\uFFFD';
+    private static final byte[] OCTETS_TO_HEX = new byte['f' + 1];
```

Comment:
I may stick with `int[]` since the difference in size is so minor, but it could give the JVM a better time of finding some optimization.

Reply:
I'm not sure. I'd prefer `byte[]` because it's more compact and thus better utilization of CPU cache.